include/layers/PoolingLayer.hpp

```diff
@@ -117,66 +196,102 @@ std::vector<ValueType> PoolingLayerImpl<ValueType>::run(
+             w < (spatial_dims > 1
+                      ? this->outputShape_[this->outputShape_.dims() -
+                                           spatial_dims + 1]
+                      : 1);
+             w++) {
+          std::vector<ValueType> pooling_buf;
+
+          int start_h =
+              static_cast<int>(h * strides_[0]) - static_cast<int>(pads_[0]);
+          int start_w = spatial_dims > 1 ? static_cast<int>(w * strides_[1]) -
+                                               static_cast<int>(pads_[2])
+                                         : 0;
+
+          for (size_t kh = 0; kh < poolingShape_[0]; kh++) {
+            for (size_t kw = 0; kw < (spatial_dims > 1 ? poolingShape_[1] : 1);
+                 kw++) {
+              int pos_h = start_h + static_cast<int>(kh * dilations_[0]);
+              int pos_w = spatial_dims > 1
+                              ? start_w + static_cast<int>(kw * dilations_[1])
+                              : 0;
+
+              if (pos_h >= 0 &&
+                  pos_h < static_cast<int>(
+                              this->inputShape_[this->inputShape_.dims() -
+                                                spatial_dims]) &&
+                  (spatial_dims <= 1 ||
+                   (pos_w >= 0 &&
+                    pos_w < static_cast<int>(
+                                this->inputShape_[this->inputShape_.dims() -
+                                                  spatial_dims + 1])))) {
+                std::vector<size_t> input_coords(this->inputShape_.dims(), 0);
+                if (batch_dim >= 0) input_coords[batch_dim] = n;
+                if (channel_dim >= 0) input_coords[channel_dim] = c;
+                input_coords[this->inputShape_.dims() - spatial_dims] = pos_h;
+                if (spatial_dims > 1)
+                  input_coords[this->inputShape_.dims() - spatial_dims + 1] =
+                      pos_w;
+
+                size_t input_index = this->inputShape_.get_index(input_coords);
+                pooling_buf.push_back(input[input_index]);
               }
             }
           }
-          switch (poolingType_) {
-            case kAverage:
-              res.push_back(avg_pooling(pooling_buf));
-              break;
-            case kMax:
-              res.push_back(max_pooling(pooling_buf));
-              break;
-            default:
-              throw std::runtime_error("Unknown pooling type");
+
+          std::vector<size_t> output_coords(this->outputShape_.dims(), 0);
+          if (batch_dim >= 0) output_coords[batch_dim] = n;
+          if (channel_dim >= 0) output_coords[channel_dim] = c;
+          output_coords[this->outputShape_.dims() - spatial_dims] = h;
+          if (spatial_dims > 1)
+            output_coords[this->outputShape_.dims() - spatial_dims + 1] = w;
+
+          size_t output_index = this->outputShape_.get_index(output_coords);
+
+          if (!pooling_buf.empty()) {
+            switch (this->poolingType_) {
```

Comment:
Add a `default:` in the `switch (poolingType_)` to guard unknown enums; same in the TBB variant.